src/System.Net.Http/src/uap/System/Net/HttpHandlerToFilter.cs

```diff
@@ -21,70 +21,249 @@
+
         private int _filterMaxVersionSet;
+        private HttpClientHandler _handler;
 
-        internal HttpHandlerToFilter(RTHttpBaseProtocolFilter filter)
+        internal HttpHandlerToFilter(
+            RTHttpBaseProtocolFilter filter,
+            HttpClientHandler handler)
         {
             if (filter == null)
             {
                 throw new ArgumentNullException(nameof(filter));
             }
 
-            _next = filter;
+            _filter = filter;
             _filterMaxVersionSet = 0;
+            _handler = handler;
+            
+            _filterWithNoCredentials = new Lazy<RTHttpBaseProtocolFilter>(InitFilterWithNoCredentials);
         }
 
         internal string RequestMessageLookupKey { get; set; }
         internal string SavedExceptionDispatchInfoLookupKey { get; set; }
 
         protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancel)
         {
+            int redirects = 0;
+            HttpMethod requestHttpMethod;
+            bool skipRequestContentIfPresent = false;
+            HttpResponseMessage response = null;
+
             if (request == null)
             {
                 throw new ArgumentNullException(nameof(request));
             }
-            cancel.ThrowIfCancellationRequested();
 
-            RTHttpRequestMessage rtRequest = await ConvertRequestAsync(request).ConfigureAwait(false);
+            requestHttpMethod = request.Method;
 
-            RTHttpResponseMessage rtResponse;
-            try
+            while (true)
             {
-                rtResponse = await _next.SendRequestAsync(rtRequest).AsTask(cancel).ConfigureAwait(false);
+                cancel.ThrowIfCancellationRequested();
+
+                if (response != null)
+                {
+                    response.Dispose();
+                    response = null;
+                }
+
+                RTHttpRequestMessage rtRequest = await ConvertRequestAsync(
+                    request,
+                    requestHttpMethod,
+                    skipRequestContentIfPresent).ConfigureAwait(false);
+
+                RTHttpResponseMessage rtResponse;
+                try
+                {
+                    if (redirects > 0)
+                    {
+                        rtResponse = await FilterWithNoCredentials.SendRequestAsync(rtRequest).AsTask(cancel).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        rtResponse = await _filter.SendRequestAsync(rtRequest).AsTask(cancel).ConfigureAwait(false);
+                    }
+                }
+                catch (TaskCanceledException)
+                {
+                    throw;
+                }
+                catch (Exception)
+                {
+                    object info;
+                    if (rtRequest.Properties.TryGetValue(SavedExceptionDispatchInfoLookupKey, out info))
```

Comment:
Nit: could use:
```C#
if (rtRequest.Properties.TryGetValue(SavedExceptionDispatchInfoLookupKey, out object info))
```
instead of declaring it on the previous line.